### PR TITLE
add some basic build tests to get started, as well as a basic test framework

### PR DESCRIPTION
**Pull request by jakemac53**
I also snuck in a needed update to `AssetReader`, which lets you list files matching a list of `InputSet`s. This was the last step needed to have a fully injectable asset system.

### CODE REVIEW

File: lib/src/asset/file_based.dart [hunk 1]
```diff
@@ -46,6 +48,35 @@ class FileBasedAssetReader implements AssetReader {
...
   }
+
+  /// Searches for all [AssetId]s matching [inputSet]s.
+  Stream<AssetId> listAssetIds(Iterable<InputSet> inputSets) async* {
+    var seenAssets = new Set<AssetId>();
+    for (var inputSet in inputSets) {
+      var packageNode = packageGraph[inputSet.package];
+      var packagePath = packageNode.location.toFilePath();
+      for (var glob in inputSet.globs) {
+        var files = glob.listSync(followLinks: false, root: packagePath).where(
```
Comment (sigmundch) [inline lib/src/asset/file_based.dart]: might not matter, but I guess you could do this async, given that you are already in an async* method?

File: test/generate/build_test.dart [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+main() {
+  group('build', () {
+    group('with root package inputs', () {
+      test('one phase, one builder, one-to-one outputs', () async {
+        var phases = [
+          [
+            new Phase([new CopyBuilder()], [new InputSet('a')]),
+          ]
+        ];
+        await testPhases(phases, {'a|web/a.txt': 'a', 'a|lib/b.txt': 'b',},
```
Comment (sigmundch) [inline test/generate/build_test.dart]: `,` :) (same below)

File: test/generate/build_test.dart [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      test('can\'t output files in non-root packages', () async {
+        var phases = [
+          [
+            new Phase([new CopyBuilder()], [new InputSet('b')]),
+          ]
+        ];
+        await testPhases(phases, {'b|lib/b.txt': 'b',},
+            outputs: {}, status: BuildStatus.Failure);
+      },
+          skip: 'Failing, InMemoryAssetWriter doesn\'t throw. Need to handle '
```
Comment (sigmundch) [inline test/generate/build_test.dart]: weird indentation from dartfmt? :-/

File: test/generate/build_test.dart [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+      ];
+      await testPhases(phases, {'a|lib/a.txt': 'a', 'a|lib/a.txt.copy': 'a'},
+          status: BuildStatus.Failure,
+          exceptionMatcher: invalidOutputException);
+    },
+        skip: 'InMemoryAssetWriter doesn\'t check this, should be handled at '
+            'the BuildStep level.');
+  });
+
+  test('outputs from builds shouldn\'t be inputs to later ones', () {},
```
Comment (sigmundch) [inline test/generate/build_test.dart]: mmm - i thought you allowed to chain build steps. Is that a no no?

File: test/generate/build_test.dart [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      remainingOutputIds.remove(asset.id);
+
+      /// Check that the writer wrote the assets
+      expect(actualAssets, contains(asset.id));
+      expect(actualAssets[asset.id], asset.stringContents);
+
+      /// Check that the assets exist in [result.outputs].
+      var actual = result.outputs
+          .firstWhere((output) => output.id == asset.id, orElse: () => null);
+      expect(actual, isNotNull);
```
Comment (sigmundch) [inline test/generate/build_test.dart]: it might be worth adding a reason here and below to make it easier to recognize why a test is failing.

Comment (sigmundch) [inline test/generate/build_test.dart]: (to print the asset id that is missing)

**Comment by sigmundch** [comment]
just minor suggestions, ow lgtm

File: lib/src/asset/file_based.dart (hunk 1 above)
Comment (jakemac53) [inline lib/src/asset/file_based.dart]: Done

File: test/generate/build_test.dart (hunk 2 above)
Comment (jakemac53) [inline test/generate/build_test.dart]: done

File: test/generate/build_test.dart (hunk 3 above)
Comment (jakemac53) [inline test/generate/build_test.dart]: ya :/

I can file a bug in a bit

File: test/generate/build_test.dart (hunk 4 above)
Comment (jakemac53) [inline test/generate/build_test.dart]: I will rename the tests, its for back to back full builds. Files created by one full build shouldn't be treated as inputs for other full builds.

File: test/generate/build_test.dart (hunk 5 above)
Comment (jakemac53) [inline test/generate/build_test.dart]: done

**Pull request merged**